### ISSUE TITLE
Aria live region queries

### ISSUE THREAD

**Issue by AriPerkkio** [testing-library/dom-testing-library#847]
### Describe the feature you'd like:

The way testing-library ensures accessibility is implemented correctly is already at really good level. However some improvements could be done to validate announcements are read correctly to assistive technologies. By announcements I mean [ARIA live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions). Currently testing these is tricky and relies highly on implementation details.

My test cases look roughly like this:

Initial state:

```html
<form>
    <label for="name">Username</label>
    <input type="text" id="name" aria-describedby="name-errors" />

    <div id="name-errors" aria-live="polite" data-testid="name-errors"></div>

    <button>Register</button>
</form>
```

Erroneous state:

```html
<form>
    <label for="name">Username</label>
    <input type="text" id="name" aria-describedby="name-errors" />

    <div id="name-errors" aria-live="polite" data-testid="name-errors">
        Username is already taken
    </div>

    <button>Register</button>
</form>
```

Test case for "visual users":

```jsx
test('user should see error if username is already taken', async () => {
    render(<UserForm />);

    // Type username of existing user
    const input = screen.getByLabelText('Username');
    userEvent.type(input, 'common-username');

    // Click register
    userEvent.click(screen.getByRole('button', { name: 'Register' }));

    // Wait for registration to finish
    const loader = screen.getByLabelText('Loading');
    await waitForElementToBeRemoved(loader);

    // API should return error describing username is already taken
    expect(input).toHaveDescription('Username is already taken');
});
```

Test case for assistive technology users:

```jsx
test('screen reader user should hear error if username is already taken', async () => {
    render(<UserForm />);

    const input = screen.getByLabelText('Username');
    const button = screen.getByRole('button', { name: 'Register' });
    const announcements = screen.getByTestId('name-errors');

    // Navigate to input
    userEvent.tab();
    expect(input).toHaveFocus();

    // Type username of existing user
    userEvent.type(document.activeElement, 'common-username');

    // Navigate to register button
    userEvent.tab();
    expect(button).toHaveFocus();

    // Announcements should be clear
    expect(announcements).toBeEmptyDOMElement();
    expect(announcements).toHaveAttribute('aria-live', 'polite');

    // Press enter to register
    fireEvent.keyDown(document.activeElement, { key: 'Enter' });

    // Wait for registration to finish
    const loader = screen.getByLabelText('Loading');
    await waitForElementToBeRemoved(loader);

    // API should return error describing username is already taken
    expect(input).toHaveDescription('Username is already taken');

    // Error should be announced to screen reader
    expect(announcements).toHaveTextContent('Username is already taken');
    expect(announcements).toHaveAttribute('aria-live', 'polite');
});
```

The explicit check for `aria-live="polite"` attribute is not very _testing-library-ish_ way of doing this.

Testing-library describes its principles with:

1. Built-in selectors find elements the way users do to help you write inclusive code
2. Tests only break when your app breaks, not implementation details

This doesn't really apply with current functionality:

1. No selector for validating users with assistive technology hear same content as the visual users are seeing
2. In the examples above the test would break if `aria-live="polite"` was replaced with `role="status"`. However users would still hear same announcements as before.

Validating these announcements is not as simple as just checking whether required `role` or `aria-live` attribute exists. Priority of announcements has to be considerd too. I'm not sure what kind of API would be ideal for this but here are some examples of expected functionality:

```html
<div role="status"></div>
<div role="alert"></div>
<div role="status"></div>
```

Expected announcements: `[]`

```html
<div role="status"></div>
<div role="alert"></div>
<div role="status">Status 2</div>
```

Expected announcements: `['Status 2']`

```html
<div role="status">Status 1</div>
<div role="alert">Alert</div>
<div role="status">Status 2</div>
```

Expected announcements: `['Status 2', 'Alert', 'Status 1']`

#### Prevent possible false positives

As developers are currently explicitly checking these `aria-live` and `role` attributes they have to consider implementation details with extreme care. Below are some examples of false positives I've encounted before. Examples are in JSX due to easier demonstration of conditional rendering.

The component renders notification messages based on given `props.notifications` array. Invalid use of `role="status"`:

```jsx
{props.notifications.length > 0 && (
    <div role='status'>{props.notifications.join(', ')}</div>
)}
```

Test case with false positive assertion:

```js
// Notifications should be announced
screen.getByRole('status', { name: 'Message 1, Message 2' });
```

The test case looks valid but the announcement is not actually made. The [specification](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA22.html) of `role="status"` states that **updates of the content** are announced. Our component is however rendering the content immediately. Instead it should update the contents of the live region.

Fixed component:

```jsx
<div role="status">
    {props.notifications.join(', ')}
</div>
```

Testing this relies highly on implementation details. 

```js
const notifications = screen.getByRole('status');

// Initially element should be rendered with empty content
expect(notification).toBeEmptyDOMElement();

// Notifications should be updated into element
await waitFor(() => {
    expect(notifications).toHaveTextContent('Message 1, Message 2');
});
```

I would like testing-library to hide such implementation detail behind a query API. There could even be an error message indicating the live region was initially rendered with content, e.g. `<div role="status">Message 1, Message 2</div> was initially rendered with content. Did you mean to update its content instead?`. 

### Suggested implementation:

I'm not expert in internals of jsdom and testing-library. Would this be possible with `MutationObserver`? Simply keep track of DOM elements which have specific `role` and `aria-live` attributes and store their content changes in global state.

I think complete support for ARIA live regions is difficult. There are so many variations that need to be considered, e.g. `aria-atomic` and `aria-relevant`.

I'm not sure what the ideal APIs would look like. Here are some ideas: 

```js
screen.getAnnouncement('Some status message');
screen.getAnnouncement('Some alert message', { assertive: true });
screen.getAnnouncement('Some alert message', { order: 1 });

const announcements = getAnnouncements();
// ['Some status message', 'Some alert message']
```

### Describe alternatives you've considered:

I haven't found any good alternatives. Currently I'm testing implementation details with testing-library. I'm not entirely happy with this approach. 

### Teachability, Documentation, Adoption, Migration Strategy:

\-

**Comment by eps1lon** [comment]
Thanks for the detailed explanation

> In the examples above the test would break if aria-live="polite" was replaced with role="status". However users would still hear same announcements as before.

This is probably better handled in `jest-dom` by adding `toHaveAriaLive('polite')` and let `toHaveAriaLive` respect implicit values. 

The whole problem sounds like a candidate for e2e testing with screen readers. Aria live regions are not that robust so even if you do have all the attributes wired up as specified you still might get different output.

> screen.getAnnouncement('Some status message');

I am currently working on automating tests with actual screen readers so this might be covered in the future with an API like 

```js
await expect(async () => {
	await page.keyboard.press("s");
}).toMatchSpeechInlineSnapshot(`
	"banner landmark"
	"Search, combo box, expanded, has auto complete, editable, Search…, blank"
`);
```

Though that would only work as an e2e test (using an actual browser) and not with `jsdom`.

The problem with your API proposal is that it isn't clear when this announcement was made. Should we permanently listen for messages and clear after every observation? What should happen with announcements that aren't asserted on?

**Comment by AriPerkkio** [comment]
> This is probably better handled in `jest-dom` by adding `toHaveAriaLive('polite')`

This would be an improvement but is still prone to false positives. Consider following case: 

JSX for easier conditional rendering demonstration again. Initial implementation:

```jsx
{error && <div role="alert">An error occured: {error}</div>}
```

```js
expect(screen.getByText("An error occured: unable to connect")).toHaveAriaLive("assertive");
```

Case above is completely valid. Announcements of `role="alert"` are done even when content of the live region is initially set. It does not require the content to be updated to an existing live region container.

Next other developer is assigned to lower the importance of these announcements from `assertive` to `polite`. Quick (faulty) implementation is to just change the `role="alert"` to `role"status"`. 

```jsx
{error && <div role="status">An error occured: {error}</div>}
```

```js
expect(screen.getByText("An error occured: unable to connect")).toHaveAriaLive("polite");
```

Test case passes but announcement is not made due to lack of content update. This case requires elements to be tracked by `MutationObserver` or similar approach. 


> The problem with your API proposal is that it isn't clear when this announcement was made. Should we permanently listen for messages and clear after every observation? What should happen with announcements that aren't asserted on?

Really good points. Clearing the announcements after each observation could work but this kind of API behaviour sounds a bit weird. 

Would it work if `screen.getAnnouncement('text')` would by default look for the latest announcement. As second argument it would take options which could be used to query previous announcements, `screen.getAnnouncement('text', { order: 2 })`. With this approach the announcement queue wouldn't be cleared ever. There is also a mention about assisitive technologies clearing the queue after assertive announcement is made. I'm not sure if this would need to be considered. [6.6 Definitions of States and Properties (all aria-* attributes) - aria-live](https://www.w3.org/TR/wai-aria/#aria-live)

It's worth to mention that this issue was earlier discussed on testing-library's Discord channel and some ideas about similar API as `jest.Mock` were thought. This would give developer the control of clearing the announcements and the latest announcement would be first in the array. I think this kind of API wouldn't be used from `screen`. It would rather be available as `import { getAnnouncements } from '@testing-library/dom'`.

At this point it's definetly clear that I'm not sure what the ideal APIs would look like. I'm more looking for open discussion and ideas about it. 


> I am currently working on automating tests with actual screen readers

This sounds really interesting! Is this open source? Can I track the process of development somewhere? I think these e2e tests would be great addition but do not replace the need for this feature.

**Comment by eps1lon** [comment]
> I think these e2e tests would be great addition but do not replace the need for this feature.

I don't see this feature ever being useful in unit tests. Or rather the amount of work would be huge. Screen-readers are already very inconsistent with announcements so tracking this behavior across screen-reader + browser combination does not seem viable to me.

Don't get me wrong, I like the idea. But without even having a picture how the automated e2e behavior would look like we might be going in the wrong direction with only using DOM APIs. 

> Can I track the process of development somewhere?

https://github.com/eps1lon/screen-reader-testing-library

**Comment by AriPerkkio** [comment]
Thanks for all the help @eps1lon. Let's close this feature request due to implementation complexity.

I'll try to implement something into my test setups to cover the conditional rendering cases mentioned earlier. It won't be perfect but should cover my use cases.

**Comment by KevinGhadyani-Okta** [comment]
What did you end up implementing? What's the trick for live regions?

**Comment by AriPerkkio** [comment]
I ended up building a library for detecting live region updates https://github.com/AriPerkkio/aria-live-capture and `extend.expect` utility for Jest and Vitest: https://github.com/AriPerkkio/extend-to-be-announced.
